File: src/main/java/qaz/code/model/ExecutionManager.java
package qaz.code.model;

import java.util.ArrayList;
import java.util.List;

public class ExecutionManager {
    public static final ExecutionManager INSTANCE = new ExecutionManager();
    private int lastExecutionCodeHash = 0;
    private int lastExecutionInputHash = 0;
    private Thread executionThread;
    private Thread killThread;

    public void process(Sheet sheet) {
        final String codeWithComments = sheet.codeProperty().get();
        final String cleanedCode = codeWithComments
                // Remove everything that isn't an instruction or newline
                .replaceAll("[^-+,.\\[\\]><\n]", "")
                // Trim start
                .replaceAll("^\s+", "")
                // Trim end
                .replaceAll("\s+$", "");
        final List<Character> input = new ArrayList<>() {{
            for (char c : sheet.inputProperty().get().toCharArray()) add(c);
        }};
        
        if (cleanedCode.hashCode() != lastExecutionCodeHash || input.hashCode() != lastExecutionInputHash) {
            lastExecutionCodeHash = cleanedCode.hashCode();
            lastExecutionInputHash = input.hashCode();
            // Stopping old threads
            if (executionThread != null && executionThread.isAlive()) executionThread.interrupt();
            if (killThread != null && killThread.isAlive()) killThread.interrupt();
            // Create new thread
            executionThread = new Thread(() -> {
                // Delay 100ms
                try {
                    Thread.sleep(100);
                }
                catch (InterruptedException ignored) {}
                // Create execution machine
                Execution execution = new Execution(Execution.Profile.DEFAULT);
                // Execute code
                Result result = execution.interpret(cleanedCode, input);
                sheet.lastResultProperty().set(result);
            });
            // And start it
            executionThread.start();
    
            // Creates parallel thread that kills execution thread if it takes too long
            killThread = new Thread(() -> {
                try {
                    Thread.sleep(5000);
                }
                catch (InterruptedException ignored) {}
                if (executionThread.isAlive()) {
                    System.err.println("Execution is taking too long, aborting...");
                    executionThread.interrupt();
                }
            });
            killThread.start();
        }
    }
}

File: src/main/java/qaz/code/model/ExecutionManager.kt
package qaz.code.model

class ExecutionManager {
    companion object {
        val INSTANCE = ExecutionManager()
        /**
         * Filters out comments, so they don't have to be processed.
         */
        val FILTER_COMMENTS = Regex("[^-+,.\\[\\]><\n]")
    }

    private var lastExecutionCodeHash = 0
    private var lastExecutionInputHash = 0
    private var executionThread: Thread? = null
    private var killThread: Thread? = null

    fun process(sheet: Sheet) {
        val codeWithComments = sheet.codeProperty.get()
        // Remove everything that isn't an instruction or newline
        val cleanedCode = codeWithComments
            .replace(FILTER_COMMENTS, "")
            .trim()

        val input = sheet.inputProperty.get().toList();

        // Only executes if the code or input has changed to prevent unnecessary executions
        if (cleanedCode.hashCode() != lastExecutionCodeHash || input.hashCode() != lastExecutionInputHash) {
            lastExecutionCodeHash = cleanedCode.hashCode()
            lastExecutionInputHash = input.hashCode()

            // Stopping old threads
            if (executionThread?.isAlive == true) executionThread!!.interrupt()
            if (killThread?.isAlive == true) killThread!!.interrupt()
            // Create new thread
            executionThread = Thread {
                // Delay 100ms
                runCatching { Thread.sleep(100) }
                // Create execution machine
                val execution = Execution(Execution.Profile.DEFAULT)
                // Execute code
                val result = execution.interpret(cleanedCode, input)
                sheet.lastResultProperty.set(result)
            }
            // And start it
            executionThread!!.start()

            // Creates parallel thread that kills execution thread if it takes too long
            killThread = Thread {
                runCatching { Thread.sleep(5000) }

                if (executionThread!!.isAlive) {
                    System.err.println("Execution is taking too long, aborting...")
                    executionThread!!.interrupt()
                }
            }
            killThread!!.start()
        }
    }
}